Add datatype iml_dataset subclassing html

knutwa-ext opened the pull request:
Add new datatype iml_dataset as a subclass of html. It is intended for use with the ImmuneML tool suite https://github.com/uio-bmi/immuneml_tools. Specifically, the purpose is to bundle files used as input and output by the ImmuneML tools. The files will be bundled as described here: 
https://planemo.readthedocs.io/en/latest/writing_advanced.html#single-html-output
See also 
https://help.galaxyproject.org/t/data-collection-not-available-as-input/5990
About ImmuneML:
https://immuneml.uio.no/

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

knutwa-ext committed:
Add datatype iml_dataset subclassing html

bgruening commented on lib/galaxy/config/sample/datatypes_conf.xml.sample:
```diff
@@ -861,6 +861,7 @@
     <datatype extension="ffindex" type="galaxy.datatypes.tabular:Tabular" display_in_upload="true" subclass="true" description_url="https://github.com/soedinglab/ffindex_soedinglab"/>
     <datatype extension="ffdata" type="galaxy.datatypes.data:Data" display_in_upload="true" subclass="true" description_url="https://github.com/soedinglab/ffindex_soedinglab"/>
     <datatype extension="pretext" type="galaxy.datatypes.binary:Pretext" display_in_upload="true" />
+    <datatype extension="iml_dataset" type="galaxy.datatypes.text:Html" subclass="True"/>
```
I would prefer a different extension here. `_dataset` seems unintuitive. Maybe `iml_summary`?

knutwa-ext replied:
We need a name to signify that it is more than a summary. I guess _dataset is sort of redundant. Maybe `immuneml_data`?

knutwa-ext replied:
To clarify: What is consumed by the next tool is a full ImmuneML dataset. And this consists of multiple files under the extra_files_path. Hence, the html serves both as a summary of the full dataset, as well as a way to pass the full dataset onwards.

bernt-matthias replied:
Then you need to define this as composite datatype (see https://docs.galaxyproject.org/en/master/dev/data_types.html#defining-basic-composite-datatypes), or?

sveinugu replied:
@bernt-matthias Yes, I agree, which is sort of the point. However due to time constraints (GCC2021 coming up quite soon...) and since HTML is a composite datatype, we decided to just subclass it for now in the `datatypes_conf.xml` file. Another argument for this solution is to make it easy for Galaxy admins to install the tools without having to update Galaxy (or the dataset package); they can instead just add a line to the `datatypes_conf.xml` file manually.

There will probably be a need to write a separate datatype class at some point, but for now HTML provides what we need. The only problem with using HTML directly is that we do not want just any HTML dataset to be selectable from the tool as input.

sveinugu replied:
@bgruening I don't think the naming of the datatype is of importance to us, so we can use whatever you think looks good. I have now asked the devs to suggest a name that more clearly describes the contents.

knutwa-ext committed:
Change datatype name to immuneml_receptors

mvdbeek committed:
Merge branch 'dev' into iml_dataset_branch

mvdbeek committed:
Allow archive mechanism for datatypes that end with html

mvdbeek reviewed (approved):
Thanks, that looks good to me. This should behave like html now, so you can use the extra_files_path to add additional files.

davelopez commented on lib/galaxy/config/sample/datatypes_conf.xml.sample:
```diff
@@ -877,6 +877,7 @@
     <datatype extension="ffindex" type="galaxy.datatypes.tabular:Tabular" display_in_upload="true" subclass="true" description_url="https://github.com/soedinglab/ffindex_soedinglab"/>
     <datatype extension="ffdata" type="galaxy.datatypes.data:Data" display_in_upload="true" subclass="true" description_url="https://github.com/soedinglab/ffindex_soedinglab"/>
     <datatype extension="pretext" type="galaxy.datatypes.binary:Pretext" display_in_upload="true" />
+    <datatype extension="immuneml_receptors.html" type="galaxy.datatypes.text:Html" subclass="True"/>
```
If I recall well `True` is not a valid boolean value in XML ;)
```suggestion
    <datatype extension="immuneml_receptors.html" type="galaxy.datatypes.text:Html" subclass="true"/>
```

davelopez committed:
Use lowercase `true` in XML

davelopez committed:
Merge branch 'dev' into iml_dataset_branch

davelopez commented on lib/galaxy/datatypes/data.py:
```diff
@@ -369,7 +369,7 @@ def to_archive(self, dataset, name=""):
         rel_paths = []
         file_paths = []
-        if dataset.datatype.composite_type or dataset.extension == 'html':
+        if dataset.datatype.composite_type or dataset.extension == '.html' or dataset.extension.endswith('.html'):
```
```suggestion
        if dataset.datatype.composite_type or dataset.extension.endswith('html'):
```
Just simplified the check a bit here since `endswith` will cover the previous check and also I've removed the preceding `.` to make sure the old check is still covered. Hope this makes sense.

davelopez reviewed (approved):
Thank you all!
Looks good to me, just a minor simplification below and I think this can be merged.

davelopez committed:
Simplify html extension check

Simplified the check a bit here since `endswith` will cover the previous check and also I've removed the preceding `.` to make sure the old check is still covered.
